README.md
## Iran Events And Holidays

A simple package to check Iran holidays and events using Jalali Calendar.

##  Installation 

You can install the package via composer:

```bash
composer require crawdance/iran-events
```


Add this line in your app.php file if your laravel version is below 5.6:

 ```sh
'providers' => [
  ...
    CrawDance\IranEvents\Providers\IranEventsServiceProvider::class,
  ...
 ];
```

## Usage:

```sh
return IranEvents::Check('1408-12-29');
```

The result is:

```sh
array:3 [▼
  "holiday" => true
  "type" => "تقویم خورشیدی"
  "desc" => "روز ملي شدن صنعت نفت ايران"
]
```

#### Hijri Events: 
```sh
return IranEvents::Check('1408-4-13');
```
```sh
array:3 [▼
  "holiday" => true
  "type" => "تقویم قمری"
  "desc" => "اربعين حسيني"
]
```

This package requires <a href="https://github.com/alkoumi/laravel-hijri-date">laravel-hijri-date</a> and <a href="https://github.com/briannesbitt/Carbon">Carbon</a> to convert Jalali inserted date to Ummul Qura Hijri and Gregorian.

## Issues

If you find a bug while using this package , please open an issue about the bug also affects ... and, unless it is obvious, the reason why you think the current result is wrong.

## License

The MIT License (MIT).


 


CHANGELOG.md
Released v1.0


src/Traits/DateCheck.php
<?php
/**
 * Developer: Keivan
 * E-mail: me@keiv.ir
 * Date:
 */
namespace CrawDance\IranEvents\Traits;

use Alkoumi\LaravelHijriDate\Hijri;
use Carbon\Carbon;
use CrawDance\IranEvents\Helpers\IranEventsRepository;
use CrawDance\IranEvents\Helpers\JalaliToGregorian;
use CrawDance\IranEvents\IranEventsInterface;

trait DateCheck
{
    public static function Check($date)
    {
        return self::check_in_events($date);
    }

    public static function convertToHijri($y, $m, $d)
    {
        $gregorian_date = self::convertToGregorian($y, $m, $d);

        $hijri = Hijri::Date('Y-m-d', $gregorian_date);
        $hijri_d = (int)explode('-', $hijri)[2];
        $hijri_m = (int)explode('-', $hijri)[1];

        return [
            'd' => $hijri_d,
            'm' => $hijri_m
        ];
    }

    public static function convertToGregorian($y, $m, $d)
    {
        try {
            $gregorian = JalaliToGregorian::convert($y, $m, $d);
            $gregorian_date = $gregorian[0].'-'.$gregorian[1].'-'.$gregorian[2] ;
            
            $gre = Carbon::createFromFormat(IranEventsInterface::FULLDATE_STRING_FORMAT, $gregorian_date.' 00:00:00')
                ->addDay(IranEventsInterface::DEFAULT_ADJUST_DAY)->format(IranEventsInterface::DEFAULT_STRING_FORMAT);
            return $gre;

        }catch(\Exceptio $exception){
            return $exception;
        }
    }

    public static function check_in_events($date)
    {
        $date = explode('-', $date);

        $persian_d = (int)$date[2];
        $persian_m = (int)$date[1];
        $persian_y = (int)$date[0];

        $hijri_date = self::convertToHijri($persian_y, $persian_m, $persian_d);

        $hijri_d = $hijri_date['d'];
        $hijri_m = $hijri_date['m'];

        $events = IranEventsRepository::getIranEvents();

        foreach ($events as $key => $value)
        {
            if ($value['Day'] == $persian_d && $value['Month'] == $persian_m && $value['Calendar'] == 'Persian')
            {
                return [
                    'holiday' => true,
                    'type' => IranEventsInterface::PERSIAN_HOLIDDAY_LABEL,
                    'desc' => $value['Title']
                ];
            }

            if ($value['Day'] == $hijri_d && $value['Month'] == $hijri_m && $value['Calendar'] == 'Hijri')
            {
                return [
                    'holiday' => true,
                    'type' => IranEventsInterface::Hijri_HOLIDDAY_LABEL,
                    'desc' => $value['Title']
                ];
            }
        }

        return [];
    }
}


src/Helpers/JalaliToGregorian.php
<?php
/**
 * Developer: Keivan
 * E-mail: me@keiv.ir
 * Date:
 */

namespace CrawDance\IranEvents\Helpers;


class JalaliToGregorian
{
    public static function convert($j_y, $j_m, $j_d, $mod = '')
    {
        $j_y = self::tr_num($j_y);
        $j_m = self::tr_num($j_m);
        $j_d = self::tr_num($j_d);
        $d_4 = ($j_y + 1) % 4;
        $doy_j = ($j_m < 7) ? (($j_m - 1) * 31) + $j_d : (($j_m - 7) * 30) + $j_d + 186;
        $d_33 = (int)((($j_y - 55) % 132) * .0305);
        $a = ($d_33 != 3 and $d_4 <= $d_33) ? 287 : 286;
        $b = (($d_33 == 1 or $d_33 == 2) and ($d_33 == $d_4 or $d_4 == 1)) ? 78 : (($d_33 == 3 and $d_4 == 0) ? 80 : 79);
        if ((int)(($j_y - 19) / 63) == 20) {
            $a--;
            $b++;
        }
        if ($doy_j <= $a) {
            $gy = $j_y + 621;
            $gd = $doy_j + $b;
        } else {
            $gy = $j_y + 622;
            $gd = $doy_j - $a;
        }
        foreach (array(0, 31, ($gy % 4 == 0) ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31) as $gm => $v) {
            if ($gd <= $v) break;
            $gd -= $v;
        }
        return ($mod == '') ? array($gy, $gm, $gd) : $gy . $mod . $gm . $mod . $gd;
    }

    public static function tr_num($str,$mod='en',$mf='٫'){
        $num_a=array('0','1','2','3','4','5','6','7','8','9','.');
        $key_a=array('۰','۱','۲','۳','۴','۵','۶','۷','۸','۹',$mf);
        return($mod=='fa')?str_replace($num_a,$key_a,$str):str_replace($key_a,$num_a,$str);
    }
}


src/Helpers/IranEventsRepository.php
<?php
/**
 * Developer: Keivan
 * E-mail: me@keiv.ir
 * Date:
 */

namespace CrawDance\IranEvents\Helpers;


class IranEventsRepository
{
    public static function getIranEvents()
    {
        return $iran_events_repo = [
            ['Title' => 'تاسوعاي حسيني', 'Calendar' => 'Hijri', 'Day' => '9', 'Month' => '1'],
            ['Title' => 'عاشوراي حسيني', 'Calendar' => 'Hijri', 'Day' => '10', 'Month' => '1'],
            ['Title' => 'اربعين حسيني', 'Calendar' => 'Hijri', 'Day' => '20', 'Month' => '2'],
            ['Title' => 'رحلت حضرت رسول اكرم صلي الله عليه و آله، 11 هـ ق - شهادت حضرت امام حسن مجتبي عليه السلام', 'Calendar' => 'Hijri', 'Day' => '28', 'Month' => '2'],
            ['Title' => 'شهادت حضرت امام رضا عليه السلام', 'Calendar' => 'Hijri', 'Day' => '30', 'Month' => '2'],
            ['Title' => 'ميلاد حضرت رسول اكرم صلي الله عليه و آله، 53 سال قبل از هجرت، روز اخلاق و مهرورزي - ميلاد حضرت امام جعفر صادق عليه‌السلام مؤسس مذهب جعفري', 'Calendar' => 'Hijri', 'Day' => '17', 'Month' => '3'],
            ['Title' => 'شهادت حضرت فاطمة زهرا سلام الله عليها', 'Calendar' => 'Hijri', 'Day' => '3', 'Month' => '6'],
            ['Title' => 'ولادت حضرت امام علي عليه السلام', 'Calendar' => 'Hijri', 'Day' => '13', 'Month' => '7'],
            ['Title' => 'مبعث حضرت رسول اكرم صلي الله عليه و آله', 'Calendar' => 'Hijri', 'Day' => '27', 'Month' => '7'],
            ['Title' => 'ولادت حضرت قائم عجل الله تعالي فرجه', 'Calendar' => 'Hijri', 'Day' => '15', 'Month' => '8'],
            ['Title' => 'شهادت حضرت علي عليه السلام', 'Calendar' => 'Hijri', 'Day' => '21', 'Month' => '9'],
            ['Title' => 'عيد سعيد فطر', 'Calendar' => 'Hijri', 'Day' => '1', 'Month' => '10'],
            ['Title' => 'شهادت حضرت امام جعفر صادق عليه السلام', 'Calendar' => 'Hijri', 'Day' => '25', 'Month' => '10'],
            ['Title' => 'عيد سعيد قربان', 'Calendar' => 'Hijri', 'Day' => '10', 'Month' => '12'],
            ['Title' => 'عيد سعيد غديرخم', 'Calendar' => 'Hijri', 'Day' => '18', 'Month' => '12'],
            ['Title' => 'آغاز نوروز', 'Calendar' => 'PersianCalendar', 'Day' => '1', 'Month' => '1'],
            ['Title' => 'عيد نوروز', 'Calendar' => 'PersianCalendar', 'Day' => '2', 'Month' => '1'],
            ['Title' => 'عيد نوروز', 'Calendar' => 'PersianCalendar', 'Day' => '3', 'Month' => '1'],
            ['Title' => 'عيد نوروز', 'Calendar' => 'PersianCalendar', 'Day' => '4', 'Month' => '1'],
            ['Title' => 'روز جمهوري اسلامي ايران', 'Calendar' => 'PersianCalendar', 'Day' => '12', 'Month' => '1'],
            ['Title' => 'روز طبيعت', 'Calendar' => 'PersianCalendar', 'Day' => '13', 'Month' => '1'],
            ['Title' => 'رحلت حضرت امام خميني -ره- رهبر كبير انقلاب و بنيان‌گذار جمهوري اسلامي ايران', 'Calendar' => 'PersianCalendar', 'Day' => '14', 'Month' => '3'],
            ['Title' => 'قيام خونين 15 خرداد', 'Calendar' => 'PersianCalendar', 'Day' => '15', 'Month' => '3'],
            ['Title' => 'پيروزي انقلاب اسلامي ايران', 'Calendar' => 'PersianCalendar', 'Day' => '22', 'Month' => '11'],
            ['Title' => 'روز ملي شدن صنعت نفت ايران', 'Calendar' => 'PersianCalendar', 'Day' => '29', 'Month' => '12'],
            ['Title' => 'تعطیلی به مناسبت عيد سعيد فطر', 'Calendar' => 'Hijri', 'Day' => '2', 'Month' => '10'],
            ['Title' => 'شهادت حضرت امام حسن عسکری عليه السلام', 'Calendar' => 'Hijri', 'Day' => '8', 'Month' => '3']
        ];
    }
}
